### ISSUE TITLE
twisted.test.test_tcp.HalfClose3TestCase.testReadNotificationRaises fails intermittently

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11212]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#1730|
|Type|defect|
|Created|2006-05-21 21:45:59Z|

At least on my x86/OS X machine.

```
  File "/Users/radix/tmp/lib/python2.4/site-packages/twisted/test/test_tcp.py", line 1124, in testReadNotificationRaises
    self.f.protocol.readConnectionLost = self.aBug
exceptions.AttributeError: 'NoneType' object has no attribute 'readConnectionLost'
```

<details><summary>Searchable metadata</summary>

```
trac-id__1730 1730
type__defect defect
reporter__radix radix
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1148247959000000 1148247959000000
changetime__1189563491000000 1189563491000000
version__ 
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#2256](https://github.com/twisted/twisted/issues/2256) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

[#2450](https://github.com/twisted/twisted/issues/2450) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This fails on Windows sometimes, too.

*bump*

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Fails in trunk on neutron (OS X host) within 100 runs pretty reliably.  In read-notification-raises-1730, ran 4000+ times without failing.  Please review!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

That looks OK, I couldn't reproduce the problem anymore under my vmware win32 machine.

Just one thing: aBug mentions readConnectionLost, it should probably mention writeConnectionLost too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [21268]) Merge read-notification-raises-1730

Author: exarkun
Reviewer: therve
Fixes [#11212](https://github.com/twisted/twisted/issues/11212)

Rewrite two unit tests for a feature of half-close support so that they
do not contain a race condition which sometimes results in the tests
failing due to setup having been only partially completed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
